rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,518 @@
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix)
+- more...
+
+###### frameworks
+
+- [dream2nix/utils](https://github.com/nix-community/dream2nix/blob/main/src/utils/config.nix)
+- [dream2nix/templates/builder](https://github.com/nix-community/dream2nix/blob/main/src/templates/builders/default.nix)
+- more... 
+
+###### Other tools
+
+Other tools that work directly with the nix AST and comments:
+
+- [noogle](https://noogle.dev) - Nix API search engine. It allows you to search functions and other expressions.
+- [nix-doc](https://github.com/lf-/nix-doc) - A Nix developer tool leveraging the rnix Nix parser for intelligent documentation search and tags generation
+- [manix](https://github.com/mlvzk/manix) - A fast CLI documentation searcher for nix. 
+
+## Proposed Improvement - Doc-blocks
+
+In general, I thought we do need two kinds of doc-strings:
+
+> The format is not final yet. Discussion may still be ongoing.
+
+#### A doc-string referencing the subsequent expression
+
+Example:
+ 
+```nix
+# somefile.nix
+
+{
+  ## <Description or Tagline>
+  ## Documentation for 'mapAttrs'
+  ## 
+  ## # Example
+  ##
+  ## <Comprehensive code>
+  ## 
+  ## # Type
+  ##    
+  ## <Type Signature>
+  mapAttrs = f: s: <...>;
+}
+```
+
+#### A Doc-string referencing the file expression
+
+Example: Uses `#|` instead of `##` to reference the whole file. It must be at the top of the file.
+ 
+```nix
+#| <Description or Tagline>
+#| 
+#| # Example
+#|
+#| <Comprehensive code>
+#| 
+#| # Type
+#|    
+#| <Type Signature>
+
+{...}:
+{
+  mapAttrs = f: s: <...>;
+}
+```
+
+# Alternatives
+[alternatives]: #alternatives
+
+In general, we need the following:
+
+1. General format for doc-strings.
+2. Format for headings and the allowed content.
+
+> It would be nice if this could be close to the markdown format.
+> Markdown is the most straightforward and most accepted format 
+> for writing and rendering documentation.
+
+
+#### Find a suitable decision in - the following matrix
+
+| General / Section | 0 `##` | 1 `/** */`   | 2 `other`   |
+|---|---|---|---|
+| 0 `# {Keyword}` | `## # Example` | `/** # Example */` |  `?` |
+| 1 `@{Keyword}:`  | `# @Example:` |`/** @Example: */` | `?` |
+| 2 `other`  | `?` | `?` | `?` |
+
+It is also possible to not implement this RFC:
+
+- By not implementing this feature, nix loses the ability for tool-generated documentation.
+- Documentation within code will remain unstable / only determined by nixdoc.
+
+## The different formats considered
+
+### `##` inspired from rust's `///` 
+
+with `Markdown` Headings
+
+Example:
+
+```nix
+# somefile.nix
+
+    ## <Description or Tagline>
+    ## 
+    ## # Example
+    ##
+    ## <Comprehensive code>
+    ## 
+    ## # Type
+    ##    
+    ## <Type Signature>
+    mapAttrs = f: s: #...
+```
+
+| Pro | Con |
+|---|---|
```

Comment:
These pro/cons are a bit hard to read. I instead recommend having a larger table with different approaches horizontally and their properties vertically, like this:

| Property/Approach | `##` | `/** */` | Javadoc | ... | 
| --- | --- | --- | --- | --- | 
| Needs termination | No | Yes | Yes ||
| Indentation clarity | Good | Poor | Poor ||
| ... | | | |

And before the table you can describe each approach and property in more detail.

Reply:
Yes! Good idea. :+1: I'll write it that way.